{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext lab_black\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dependencies    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. To visualize scRNA-seq expression data, <b>MAGIC</b> (https://www.krishnaswamylab.org/projects/magic) is used to impute the Missing data.\n",
    "2. The web-based function is backended by <b>Dash</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import magic\n",
    "import seaborn as sns\n",
    "import ipywidgets as widgets\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json, random, os\n",
    "from jupyter_dash import JupyterDash\n",
    "import dash_html_components as html\n",
    "import dash_core_components as dcc\n",
    "import dash_table\n",
    "from dash.dependencies import Input, Output, State, ALL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scRNA-seq Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This example inspects the expression profile about a gene of interest in different cell types by tissues. It uses mouse cell altas data from _Han X, Wang R, Zhou Y, Fei L et al. Mapping the Mouse Cell Atlas by Microwell-Seq. Cell 2018 Feb 22;172(5):1091-1107.e17. PMID: 29474909_.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene-cell matrix\n",
    "tissue_data_path = {\n",
    "    \"Small-Intestine\": [\n",
    "        \"data/GSM2906468_SmallIntestine1_dge.txt.gz\",\n",
    "        \"data/GSM2906469_SmallIntestine2_dge.txt.gz\",\n",
    "        \"data/GSM2906470_SmallIntestine3_dge.txt.gz\",\n",
    "    ],\n",
    "    \"Spleen\": [\"data/GSM2906471_Spleen_dge.txt.gz\"],\n",
    "    \"Kidney\": [\n",
    "        \"data/GSM2906425_Kidney1_dge.txt.gz\",\n",
    "        \"data/GSM2906426_Kidney2_dge.txt.gz\",\n",
    "    ],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# barcode -> tissue mapping\n",
    "sc_labels = pd.read_csv(\"data/MCA_CellAssignments.csv\")\n",
    "sc_labels.Tissue.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Cell.name</th>\n",
       "      <th>ClusterID</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Cell.Barcode</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Bladder_1.AAAACGAAAACGGGGCGA</td>\n",
       "      <td>Bladder_1</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>Bladder_1</td>\n",
       "      <td>AAAACGAAAACGGGGCGA</td>\n",
       "      <td>Stromal cell_Dpt high(Bladder)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Bladder_1.AAAACGAAGCGGCCGCTA</td>\n",
       "      <td>Bladder_5</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>Bladder_1</td>\n",
       "      <td>AAAACGAAGCGGCCGCTA</td>\n",
       "      <td>Stromal cell_Car3 high(Bladder)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Bladder_1.AAAACGAAGTACTAGCAT</td>\n",
       "      <td>Bladder_16</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>Bladder_1</td>\n",
       "      <td>AAAACGAAGTACTAGCAT</td>\n",
       "      <td>Vascular smooth muscle progenitor cell(Bladder)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Bladder_1.AAAACGACGTTGCTGTGT</td>\n",
       "      <td>Bladder_8</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>Bladder_1</td>\n",
       "      <td>AAAACGACGTTGCTGTGT</td>\n",
       "      <td>Vascular endothelial cell(Bladder)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Bladder_1.AAAACGAGCGAGCGAGTA</td>\n",
       "      <td>Bladder_4</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>Bladder_1</td>\n",
       "      <td>AAAACGAGCGAGCGAGTA</td>\n",
       "      <td>Urothelium(Bladder)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                     Cell.name   ClusterID   Tissue      Batch  \\\n",
       "0           1  Bladder_1.AAAACGAAAACGGGGCGA   Bladder_1  Bladder  Bladder_1   \n",
       "1           2  Bladder_1.AAAACGAAGCGGCCGCTA   Bladder_5  Bladder  Bladder_1   \n",
       "2           3  Bladder_1.AAAACGAAGTACTAGCAT  Bladder_16  Bladder  Bladder_1   \n",
       "3           4  Bladder_1.AAAACGACGTTGCTGTGT   Bladder_8  Bladder  Bladder_1   \n",
       "4           5  Bladder_1.AAAACGAGCGAGCGAGTA   Bladder_4  Bladder  Bladder_1   \n",
       "\n",
       "         Cell.Barcode                                       Annotation  \n",
       "0  AAAACGAAAACGGGGCGA                   Stromal cell_Dpt high(Bladder)  \n",
       "1  AAAACGAAGCGGCCGCTA                  Stromal cell_Car3 high(Bladder)  \n",
       "2  AAAACGAAGTACTAGCAT  Vascular smooth muscle progenitor cell(Bladder)  \n",
       "3  AAAACGACGTTGCTGTGT               Vascular endothelial cell(Bladder)  \n",
       "4  AAAACGAGCGAGCGAGTA                              Urothelium(Bladder)  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Annotation column -> cell type\n",
    "sc_labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract single cell gene expression data from the specified tissue\n",
    "# MAGIC were used to impute data\n",
    "def tissue_sc_profile(sc_labels, tissue, genes, data_path):\n",
    "    # tissue data; Parse cell type info\n",
    "    tissue_sc_config = sc_labels.loc[sc_labels.Tissue == tissue].copy()\n",
    "    tissue_sc_config.loc[:, \"Annotation\"] = tissue_sc_config.Annotation.apply(\n",
    "        lambda x: x.split(\"(\")[0]\n",
    "    ).values\n",
    "\n",
    "    magic_imputed_gene = pd.DataFrame()\n",
    "    for i, f in enumerate(data_path):\n",
    "        raw = pd.read_csv(f, sep=\" \")\n",
    "        print(\"*****\", f.split(\"/\")[-1], \"has\", raw.shape[1], \"cells\")\n",
    "        # magic impute gene expression after normalization and filter\n",
    "        upm_mtx = raw * 1e6 / raw.sum()\n",
    "        mask_low_gene = np.sum(raw.values != 0, axis=1) >= 5\n",
    "        mask_low_sqrt_mat = np.sqrt(upm_mtx.loc[mask_low_gene, :])\n",
    "        magic_op = magic.MAGIC()\n",
    "        magic_imputed = magic_op.fit_transform(mask_low_sqrt_mat.T, genes=genes)\n",
    "        magic_imputed_gene = magic_imputed_gene.append(np.power(magic_imputed, 2))\n",
    "\n",
    "    gene_exp_df = magic_imputed_gene.copy()\n",
    "\n",
    "    # only takes the single cells with labels\n",
    "    gene_exp_df = tissue_sc_config.merge(\n",
    "        gene_exp_df, left_on=\"Cell.name\", right_index=True, how=\"inner\"\n",
    "    )\n",
    "\n",
    "    # support multiple genes in the future\n",
    "    gene_exp_df[\"log2TPM\"] = np.log2(gene_exp_df[genes[0]].values + 1)\n",
    "\n",
    "    # cell type mean\n",
    "    data = []\n",
    "    for clt, grp in gene_exp_df.groupby(by=\"Annotation\"):\n",
    "        data.append([clt, len(grp), grp[genes[0]].mean()])\n",
    "\n",
    "    cell_type_mean = pd.DataFrame(\n",
    "        {\n",
    "            \"CellType\": [content[0] for content in data],\n",
    "            \"CellCount\": [content[1] for content in data],\n",
    "            \"MeanExp\": [round(content[2], 4) for content in data],\n",
    "        },\n",
    "        index=[content[0] for content in data],\n",
    "    )\n",
    "\n",
    "    return cell_type_mean, gene_exp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_tissue_profile_dashboard(sc_labels, tissue, genes, data_path):\n",
    "    cell_type_mean, gene_exp_df = tissue_sc_profile(sc_labels, tissue, genes, data_path)\n",
    "    cell_types = list(cell_type_mean[\"CellType\"].values)\n",
    "    cell_type_options = [{\"label\": ct, \"value\": ct} for ct in cell_types]\n",
    "    # fig_file = make_plot(icosl_exp_df, cell_type_mean, tissue)\n",
    "\n",
    "    return (\n",
    "        cell_type_mean,\n",
    "        # app layout: horizontal violin plot align to cell type mean table which can serve as label\n",
    "        html.Div(\n",
    "            [\n",
    "                html.Div(\n",
    "                    html.Img(id=f\"{tissue}_plot\"),\n",
    "                    style={\n",
    "                        \"display\": \"inline-block\",\n",
    "                        \"margin-top\": \"4px\",\n",
    "                    },\n",
    "                ),\n",
    "                dcc.Store(\n",
    "                    id=f\"{tissue}_curr_celltype_order\", data=json.dumps(cell_types)\n",
    "                ),\n",
    "                dcc.Store(f\"{tissue}_tag\", data=json.dumps(tissue)),\n",
    "                html.Div(\n",
    "                    dash_table.DataTable(\n",
    "                        id=f\"{tissue}_expression_table\",\n",
    "                        columns=[\n",
    "                            {\n",
    "                                \"name\": \"CellType\",\n",
    "                                \"id\": \"CellType\",\n",
    "                                \"presentation\": \"dropdown\",\n",
    "                            },\n",
    "                            {\"name\": \"CellCount\", \"id\": \"CellCount\"},\n",
    "                            {\"name\": \"MeanExp\", \"id\": \"MeanExp\"},\n",
    "                        ],\n",
    "                        data=cell_type_mean.to_dict(\"records\"),\n",
    "                        style_cell={\"minHeight\": 0, \"maxHeight\": 95, \"height\": 36.1},\n",
    "                        style_cell_conditional=[\n",
    "                            {\"if\": {\"column_id\": \"CellType\"}, \"textAlign\": \"left\"}\n",
    "                        ],\n",
    "                        editable=True,\n",
    "                        dropdown={\n",
    "                            \"CellType\": {\"options\": cell_type_options},\n",
    "                        },\n",
    "                    ),\n",
    "                    style={\n",
    "                        \"vertical-align\": \"top\",\n",
    "                        \"display\": \"inline-block\",\n",
    "                    },\n",
    "                ),\n",
    "            ],\n",
    "            style={\n",
    "                \"display\": \"block\",\n",
    "                \"margin-left\": \"23%\",\n",
    "                \"margin-right\": \"auto\",\n",
    "            },\n",
    "        ),\n",
    "        gene_exp_df,\n",
    "    )\n",
    "\n",
    "\n",
    "def make_plot(data, df, tissue):\n",
    "    orders = list(df[\"CellType\"].values)\n",
    "    row_size = 13.05 / 28\n",
    "    fig, ax = plt.subplots(figsize=(6, row_size * len(orders)), dpi=100)\n",
    "    sns.violinplot(\n",
    "        x=\"log2TPM\",\n",
    "        y=\"Annotation\",\n",
    "        data=data,\n",
    "        order=orders,\n",
    "        scale=\"width\",\n",
    "        inner=None,\n",
    "        linewidth=1,\n",
    "        cut=0,\n",
    "        ax=ax,\n",
    "    )\n",
    "    ax.tick_params(\n",
    "        axis=\"x\",  # changes apply to the x-axis\n",
    "        which=\"both\",  # both major and minor ticks are affected\n",
    "        bottom=False,  # ticks along the bottom edge are off\n",
    "        top=True,  # ticks along the top edge are off\n",
    "        labelbottom=False,\n",
    "        labeltop=True,\n",
    "    )\n",
    "    ax.tick_params(\n",
    "        axis=\"y\",  # changes apply to the x-axis\n",
    "        which=\"both\",  # both major and minor ticks are affected\n",
    "        bottom=False,  # ticks along the bottom edge are off\n",
    "        top=False,  # ticks along the top edge are off\n",
    "        right=True,\n",
    "        left=False,\n",
    "        labelleft=False,\n",
    "    )\n",
    "    ax.invert_xaxis()\n",
    "    ax.xaxis.label.set_visible(False)\n",
    "    ax.yaxis.label.set_visible(False)\n",
    "    out_file = f\"assets/{tissue}_{random.randrange(10000)}.png\"\n",
    "    fig.savefig(out_file, bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    return out_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** GSM2906468_SmallIntestine1_dge.txt.gz has 4764 cells\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 4764 cells and 14570 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 15.55 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 2.63 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 2.80 seconds.\n",
      "  Calculated graph and diffusion operator in 23.06 seconds.\n",
      "  Calculating imputation...\n",
      "Calculated MAGIC in 23.74 seconds.\n",
      "***** GSM2906469_SmallIntestine2_dge.txt.gz has 2215 cells\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 2215 cells and 12295 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 7.18 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 0.51 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 0.50 seconds.\n",
      "  Calculated graph and diffusion operator in 9.01 seconds.\n",
      "  Calculating imputation...\n",
      "Calculated MAGIC in 9.20 seconds.\n",
      "***** GSM2906470_SmallIntestine3_dge.txt.gz has 2005 cells\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 2005 cells and 11781 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 5.99 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 0.43 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 0.46 seconds.\n",
      "  Calculated graph and diffusion operator in 7.31 seconds.\n",
      "  Calculating imputation...\n",
      "Calculated MAGIC in 7.45 seconds.\n",
      "***** GSM2906471_Spleen_dge.txt.gz has 2417 cells\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 2417 cells and 11767 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 7.62 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 0.81 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 0.78 seconds.\n",
      "  Calculated graph and diffusion operator in 10.05 seconds.\n",
      "  Calculating imputation...\n",
      "Calculated MAGIC in 10.20 seconds.\n",
      "***** GSM2906425_Kidney1_dge.txt.gz has 445 cells\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 445 cells and 9726 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 1.44 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 0.06 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 0.04 seconds.\n",
      "  Calculated graph and diffusion operator in 1.57 seconds.\n",
      "  Calculating imputation...\n",
      "Calculated MAGIC in 1.72 seconds.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/coco/miniconda3/envs/scrna/lib/python3.7/site-packages/magic/magic.py:486: UserWarning: genes ['Icosl'] missing from input data\n",
      "  genes[~np.isin(genes, gene_names)]\n",
      "/home/coco/miniconda3/envs/scrna/lib/python3.7/site-packages/scprep/select.py:454: UserWarning: Selecting 0 columns.\n",
      "  warnings.warn(\"Selecting 0 columns.\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** GSM2906426_Kidney2_dge.txt.gz has 6220 cells\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 6220 cells and 14298 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 20.23 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 5.19 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 5.17 seconds.\n",
      "  Calculated graph and diffusion operator in 32.96 seconds.\n",
      "  Calculating imputation...\n",
      "Calculated MAGIC in 33.38 seconds.\n"
     ]
    }
   ],
   "source": [
    "# create layouts for each of the tissue\n",
    "tissue_layout = {\n",
    "    tissue: create_tissue_profile_dashboard(sc_labels, tissue, [\"Icosl\"], data_path)\n",
    "    for tissue, data_path in tissue_data_path.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "app = JupyterDash(__name__)\n",
    "app.layout = html.Div(\n",
    "    [\n",
    "        html.H1(\n",
    "            \"Rearrange Cell Type Orders in Violin Plots\",\n",
    "            style={\"text-align\": \"center\"},\n",
    "        ),\n",
    "        # organize tissues' dashboards into tabs\n",
    "        dcc.Tabs(\n",
    "            id=\"tissue_selector\",\n",
    "            value=\"Small-Intestine\",\n",
    "            children=[\n",
    "                dcc.Tab(\n",
    "                    label=f\"{tissue}_tab\",\n",
    "                    value=tissue,\n",
    "                    children=[\n",
    "                        html.H3(\n",
    "                            'Change the cell type in \"CellType\" column to reorder',\n",
    "                            style={\"text-align\": \"center\"},\n",
    "                        ),\n",
    "                        data[1],\n",
    "                        html.Div(\n",
    "                            [\n",
    "                                html.H3(\n",
    "                                    \"Horizontal View\",\n",
    "                                    style={\"text-align\": \"center\"},\n",
    "                                ),\n",
    "                                html.Button(\n",
    "                                    \"Download Figure\",\n",
    "                                    id=f\"{tissue}_btn_figure\",\n",
    "                                    style={\n",
    "                                        \"display\": \"block\",\n",
    "                                        \"margin-left\": \"auto\",\n",
    "                                        \"margin-right\": \"auto\",\n",
    "                                        \"margin-bottom\": \"10px\",\n",
    "                                    },\n",
    "                                ),\n",
    "                                dcc.Download(id=f\"{tissue}_download_figure\"),\n",
    "                                html.Img(\n",
    "                                    id=f\"{tissue}_final_plot\",\n",
    "                                    style={\n",
    "                                        \"width\": \"70%\",\n",
    "                                        \"height\": \"auto\",\n",
    "                                        \"display\": \"block\",\n",
    "                                        \"margin-left\": \"auto\",\n",
    "                                        \"margin-right\": \"auto\",\n",
    "                                    },\n",
    "                                ),\n",
    "                                html.P(id=f\"{tissue}_display_cell_type_order\"),\n",
    "                            ],\n",
    "                        ),\n",
    "                    ],\n",
    "                )\n",
    "                for tissue, data in tissue_layout.items()\n",
    "            ],\n",
    "            style={\"padding\": \"10px\"},\n",
    "        ),\n",
    "    ]\n",
    ")\n",
    "\n",
    "# arrange row order in the table and generating reordered violine plot\n",
    "for tissue in tissue_layout.keys():\n",
    "\n",
    "    @app.callback(\n",
    "        Output(f\"{tissue}_expression_table\", \"data\"),\n",
    "        Output(f\"{tissue}_curr_celltype_order\", \"data\"),\n",
    "        Input(f\"{tissue}_expression_table\", \"data\"),\n",
    "        Input(f\"{tissue}_curr_celltype_order\", \"data\"),\n",
    "        State(f\"{tissue}_tag\", \"data\"),\n",
    "        prevent_initial_call=True,\n",
    "    )\n",
    "    def swap_cell_type_order(df_dict, curr_ctype_order_json, tissue_tag):\n",
    "        selected_tissue = json.loads(tissue_tag)\n",
    "        cell_type_mean = tissue_layout[selected_tissue][0]\n",
    "        cell_types = list(cell_type_mean[\"CellType\"].values)\n",
    "        current_ctype_order = json.loads(curr_ctype_order_json)\n",
    "\n",
    "        df = pd.DataFrame.from_dict(df_dict)\n",
    "        values = df[\"CellType\"].values\n",
    "\n",
    "        for ori, modified in zip(current_ctype_order, values):\n",
    "            if ori != modified:\n",
    "                break\n",
    "\n",
    "        values[current_ctype_order.index(modified)] = ori\n",
    "        return cell_type_mean.loc[values, :].to_dict(\"records\"), json.dumps(\n",
    "            list(values)\n",
    "        )\n",
    "\n",
    "    @app.callback(\n",
    "        Output(f\"{tissue}_plot\", \"src\"),\n",
    "        Input(f\"{tissue}_expression_table\", \"data\"),\n",
    "        Input(f\"{tissue}_plot\", \"src\"),\n",
    "        State(f\"{tissue}_tag\", \"data\"),\n",
    "    )\n",
    "    def update_plot(df_dict, old_plot, tissue_tag):\n",
    "        selected_tissue = json.loads(tissue_tag)\n",
    "        if old_plot is not None:\n",
    "            os.remove(old_plot)\n",
    "        df = pd.DataFrame.from_dict(df_dict)\n",
    "\n",
    "        return make_plot(tissue_layout[selected_tissue][2], df, selected_tissue)\n",
    "\n",
    "    @app.callback(\n",
    "        Output(f\"{tissue}_display_cell_type_order\", \"children\"),\n",
    "        Output(f\"{tissue}_final_plot\", \"src\"),\n",
    "        Input(f\"{tissue}_expression_table\", \"data\"),\n",
    "        Input(f\"{tissue}_final_plot\", \"src\"),\n",
    "        State(f\"{tissue}_tag\", \"data\"),\n",
    "    )\n",
    "    def update_final_plot(df_dict, old_plot, tissue_tag):\n",
    "        selected_tissue = json.loads(tissue_tag)\n",
    "        if old_plot is not None:\n",
    "            os.remove(old_plot)\n",
    "        df = pd.DataFrame.from_dict(df_dict)\n",
    "\n",
    "        orders = list(df[\"CellType\"].values)\n",
    "        fig, ax = plt.subplots(figsize=(20, 6), dpi=100)\n",
    "        sns.violinplot(\n",
    "            x=\"Annotation\",\n",
    "            y=\"log2TPM\",\n",
    "            data=tissue_layout[selected_tissue][2],\n",
    "            order=orders,\n",
    "            scale=\"width\",\n",
    "            inner=None,\n",
    "            linewidth=1,\n",
    "            cut=0,\n",
    "            legend=True,\n",
    "            ax=ax,\n",
    "        )\n",
    "        ax.tick_params(axis=\"x\", rotation=90)  # changes apply to the x-axis\n",
    "        out_file = f\"assets/{selected_tissue}_final_{random.randrange(10000)}.png\"\n",
    "        fig.savefig(out_file, bbox_inches=\"tight\")\n",
    "        plt.close()\n",
    "        return \", \".join(orders), out_file\n",
    "\n",
    "    @app.callback(\n",
    "        Output(f\"{tissue}_download_figure\", \"data\"),\n",
    "        Input(f\"{tissue}_btn_figure\", \"n_clicks\"),\n",
    "        State(f\"{tissue}_expression_table\", \"data\"),\n",
    "        State(f\"{tissue}_tag\", \"data\"),\n",
    "        prevent_initial_call=True,\n",
    "    )\n",
    "    def make_figure(n_clicks, df_dict, tissue_tag):\n",
    "        selected_tissue = json.loads(tissue_tag)\n",
    "        df = pd.DataFrame.from_dict(df_dict)\n",
    "        orders = list(df[\"CellType\"].values)\n",
    "        fig, ax = plt.subplots(figsize=(20, 6), dpi=400)\n",
    "        sns.violinplot(\n",
    "            x=\"Annotation\",\n",
    "            y=\"log2TPM\",\n",
    "            data=tissue_layout[selected_tissue][2],\n",
    "            order=orders,\n",
    "            scale=\"width\",\n",
    "            inner=None,\n",
    "            linewidth=1,\n",
    "            cut=0,\n",
    "            legend=True,\n",
    "            ax=ax,\n",
    "        )\n",
    "        ax.tick_params(axis=\"x\", rotation=90)  # changes apply to the x-axis\n",
    "        out_file = f\"assets/{selected_tissue}_final_figure.png\"\n",
    "        fig.savefig(out_file, bbox_inches=\"tight\")\n",
    "        plt.close()\n",
    "\n",
    "        with open(f\"assets/{selected_tissue}_final_figure_order.txt\", \"w\") as o:\n",
    "            o.write(\", \".join(orders))\n",
    "\n",
    "        return dcc.send_file(out_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "192.168.0.3 - - [10/Jul/2021 12:37:47] \"\u001b[37mGET /_shutdown_96511a27-5262-40ab-acdc-495cf3464116 HTTP/1.1\u001b[0m\" 200 -\n",
      " * Running on http://192.168.0.3:8892/ (Press CTRL+C to quit)\n",
      "192.168.0.3 - - [10/Jul/2021 12:37:47] \"\u001b[37mGET /_alive_96511a27-5262-40ab-acdc-495cf3464116 HTTP/1.1\u001b[0m\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash app running on http://192.168.0.3:8892/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "192.168.0.5 - - [10/Jul/2021 12:37:50] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:50] \"\u001b[37mGET /_dash-dependencies HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:50] \"\u001b[37mGET /_dash-layout HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:50] \"\u001b[37mGET /_dash-component-suites/dash_table/async-highlight.js HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:50] \"\u001b[37mGET /_dash-component-suites/dash_table/async-table.js HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:52] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:53] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:53] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:53] \"\u001b[37mGET /assets/Small-Intestine_final_696.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:53] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:53] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:54] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:37:54] \"\u001b[37mGET /assets/Small-Intestine_2937.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:12] \"\u001b[37mGET /assets/Spleen_final_1981.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:12] \"\u001b[37mGET /assets/Spleen_3076.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:17] \"\u001b[37mGET /assets/Kidney_final_1834.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:17] \"\u001b[37mGET /assets/Kidney_5739.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:34] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:36] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:36] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:36] \"\u001b[37mGET /assets/Spleen_final_7332.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:36] \"\u001b[37mGET /assets/Spleen_865.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:39] \"\u001b[35m\u001b[1mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 500 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:40] \"\u001b[35m\u001b[1mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 500 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:40] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:40] \"\u001b[37mGET /assets/Spleen_1165.png HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:45] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:45] \"\u001b[35m\u001b[1mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 500 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:45] \"\u001b[37mPOST /_dash-update-component HTTP/1.1\u001b[0m\" 200 -\n",
      "192.168.0.5 - - [10/Jul/2021 12:38:45] \"\u001b[37mGET /assets/Spleen_2346.png HTTP/1.1\u001b[0m\" 200 -\n"
     ]
    }
   ],
   "source": [
    "app.run_server(host=\"192.168.0.3\", port=8892, debug=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scrna",
   "language": "python",
   "name": "scrna"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
